adafruit_connection_manager.py

```diff
@@ -253,64 +286,63 @@ def get_socket(
             host, port, 0, self._socket_pool.SOCK_STREAM
         )[0]
 
-        try_count = 0
-        socket = None
-        last_exc = None
-        while try_count < 2 and socket is None:
-            try_count += 1
-            if try_count > 1:
-                if any(
-                    socket
-                    for socket, free in self._available_socket.items()
-                    if free is True
-                ):
-                    self._free_sockets()
-                else:
-                    break
+        result = self._get_connected_socket(
+            addr_info, host, port, timeout, is_ssl, ssl_context
+        )
+        if isinstance(result, Exception):
+            # Got an error, if there are any available sockets, free them and try again
+            if self.available_socket_count:
+                self._free_sockets()
+                result = self._get_connected_socket(
+                    addr_info, host, port, timeout, is_ssl, ssl_context
+                )
+        if isinstance(result, Exception):
+            raise RuntimeError(f"Error connecting socket: {result}") from result
```

Comment:
@RetiredWizard this is tough. I wouldn't want a core library printing. I think with a little thought, we could flag a debug mode that prints them all (maybe with a `connection_manager_set_debug(socket)`. I also have a fear on too much bloat...

What about:
```py
        last_result = ""
        result = self._get_connected_socket(
            addr_info, host, port, timeout, is_ssl, ssl_context
        )
        if isinstance(result, Exception):
            # Got an error, if there are any available sockets, free them and try again
            if self.available_socket_count:
                last_result = f", first error: {result}"
                self._free_sockets()
                result = self._get_connected_socket(
                    addr_info, host, port, timeout, is_ssl, ssl_context
                )
        if isinstance(result, Exception):
            raise RuntimeError(f"Error connecting socket: {result}{last_result}") from result
```

Reply:
This would update this test:
```py

def test_get_socket_runtime_error_ties_again_only_once():
    mock_pool = mocket.MocketPool()
    mock_socket_1 = mocket.Mocket()
    mock_socket_2 = mocket.Mocket()
    mock_pool.socket.side_effect = [
        mock_socket_1,
        RuntimeError("error 1"),
        RuntimeError("error 2"),
        RuntimeError("error 3"),
        mock_socket_2,
    ]

    free_sockets_mock = mock.Mock()
    connection_manager = adafruit_connection_manager.ConnectionManager(mock_pool)
    connection_manager._free_sockets = free_sockets_mock

    # get a socket and then mark as free
    socket = connection_manager.get_socket(mocket.MOCK_HOST_1, 80, "http:")
    assert socket == mock_socket_1
    connection_manager.free_socket(socket)
    free_sockets_mock.assert_not_called()

    # try to get a socket that returns a RuntimeError twice
    with pytest.raises(RuntimeError) as context:
        connection_manager.get_socket(mocket.MOCK_HOST_2, 80, "http:")
    assert "Error connecting socket: error 2, first error: error 1" in str(context)
    free_sockets_mock.assert_called_once()
```

Reply:
Yea If that does what it looks like it does, that's great. I also see the argument to drop for the sake of bloat though. Some of the WiFi boards have really thin available memory and a few bytes can make a big difference. I'll let you make the call :grin:

Reply:
@dhalbert are you good with this change?

Reply:
this seems ok. You can save just save the exception (`first_exception`) and then only make a formatted string in the second `if`.

Yes, reducing code size is good wherever you can do it for frozen libraries.